IPLD-Data-Model-v1.md

```diff
@@ -0,0 +1,18 @@
+# [WIP] IPLD Data Model
+
+## Simple Types
+
+* Boolean
+* Null
+* String
+* Integer
+  * Must support BigInt
```

Comment:
I'd rather avoid promising bigint *as a core type* and revisit that in more detail later.

Reply:
This was something Juan was pretty adamant about. We don't want to lose precision when mutating a graph from a language or codec that didn't implement the type completely.

Reply:
:grimacing:

Reply:
I'm just gonna come out and say right now I'm uncomfortable promising to actually implement that.  Computers tend to have this weird hangup about numbers being made out of countable number of bits.  Fighting it tends to not result in code that's fast or APIs that anyone is happy about.

I'm afraid if we commit to this, then we're going to end up with implementations that ignore this part of the standard *anyway* and those will become the more popular ones, because nobody likes to program with `MySpecialBigInt` type wrappers rather than their language's native `int`.  And people will constantly cast things to their language's `int` anyway when doing operations.

We can and should have some definitions of BigInt that we support well up at the "Types" level, higher.  But I'm extremely dubious that doing it at the *Data Model* layer is even viable.

Reply:
It's interesting to consider here what guarantees or requirements we can actually make at this spec level.

I'm comfortable saying that we can make requirements of the *format*. But when it comes to the **representation** codecs give in each language there's a limit to the guarantees we can make. Obviously codecs should make every effort possible to support the type but it may not always be possible.

For `dag-json` in JavaScript I'm confident I can support this even though the language is not particularly helpful. JSON already supports BigInt theoretically, it's JavaScript's number type that is problematic. But since I own the `dag-json` serializer/deserializer I can look for BigInt types and for serialized numbers that are too large and represent them with BigInt types rather than regular numbers types.

Reply:
I'm just not convinced that this either A) solves real problems B) in *a way that people will actually use, for human and API usability reasons* and C) is worth the effort in making thoroughly nonstandard parsers for any of the underlying serial representations that we use which are already specified and lack this capability (e.g. all of them).

I definitely buy that we can write custom parsers that keep (for example) reading another base ten ascii character in JSON and in its internal model racks up a value that's well beyond the 4 billion range.

I *don't* buy that it's plausibly usable to then store that number in some blessed-at-the-Data-Model-level BigInt type that's exposed to the user and which will rule all BigInt types.  The prior art for BigInt types is large; there's many serial implementations of them; and I don't think we should try to wade into that holy war at the Data Model level.  (The "Types" level, above?  Sure.  But not the Data Model, where we'll be stuck with it in perpetuity and have limited options supporting variations.)

I don't buy that we'll have very friendly APIs if any integer is implicitly `Either<int,BigInt>` -- that might fly in a fairly untyped language like javascript, but it'll still produce utterly bizarre runtime errors when users actually encounter it; and it's not going to fly well *at all* in more strongly typed languages.  In particular, this will yield a fractal of trouble if we make other choices like allowing numeric map keys; `map[Either<int,BigInt>]interface{}` is :nauseated_face:   I don't think we've specified whether numeric map keys are allowed yet, so we certainly can make other choices to avoid this problem... but I think it's critical to make sure we've fully considered these kinds of kinds of choices before committing to a definition of numerics will require us to allow bigints anywhere we allow ints.

I don't buy that IPLD having native BigInt everywhere is either necessary nor (more importantly) *sufficient* for making lossless transformations of IPLD data the norm.  As long as people write code in languages that don't have native BigInt, they'll be writing bindings of IPLD to types like `struct{ x int }`, doing their logical transformations on those structures, and emitting IPLD again.  In such usage, it doesn't matter in the slightest if we've gone through tons of work to make native bigints; the least expansive primitive types used anywhere in the transformation code are going to be the dominant factor; and for the forseeable future that means anyone designing a durable protocol that works smoothly in many modern languages is still going to want to consciously constrain themselves to int64 or int32.  (http://reasonablypolymorphic.com/blog/protos-are-wrong/ has some commentary similar to this; namely, it's rare in practice to see programs which inspect only one field out of a protobuf type; the vast majority of programs that people write will load, transform in that language's types, perform some computation, and emit again; this is the path that we should concern ourselves with.)  So, I would argue that the necessary thing for lossless transformations of IPLD data is more to do with the implementation and *having a language* that uses bigints for its basic numeric types than anything else.

---

tl;dr; I think we'd be better off saying "implementations SHOULD support the largest integer ranges possible and MUST support at least int32 ranges; protocols SHOULD NOT attempt to use any kind of "INT_MAX" values as sentinels"... and trying to solve infinite portability problems at the level where they're actually most likely to be problematic: transformation code and the languages its written in.